src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Measurement.cs

```diff
@@ -76,6 +78,63 @@ public Measurement(T value, params ReadOnlySpan<KeyValuePair<string, object?>> t
         public T Value { get; }
 
         // Private helper to copy IEnumerable to array. We have it to avoid adding dependencies on System.Linq
-        private static KeyValuePair<string, object?>[] ToArray(IEnumerable<KeyValuePair<string, object?>>? tags) => tags is null ? Instrument.EmptyTags : new List<KeyValuePair<string, object?>>(tags).ToArray();
+        private static KeyValuePair<string, object?>[] ToArray(IEnumerable<KeyValuePair<string, object?>>? tags)
+        {
+            if (tags is null)
+                return Instrument.EmptyTags;
+
+            // When the input is a collection, we can allocate a correctly sized array and copy directly into it.
+            if (tags is ICollection<KeyValuePair<string, object?>> collection)
+            {
+                int items = collection.Count;
+
+                if (items == 0)
+                    return Instrument.EmptyTags;
+
+                KeyValuePair<string, object?>[] result = new KeyValuePair<string, object?>[items];
+                collection.CopyTo(result, 0);
+                return result;
+            }
+
+            // In any other case, we must enumerate the input.
+            // We use a pooled array as a buffer to avoid allocating until we know the final size we need.
+            // This assumes that there are 32 or fewer tags, which is a reasonable assumption for most cases.
+            // In the worst case, we will grow the buffer by renting a larger array.
+            int count = 0;
+            KeyValuePair<string, object?>[] array = ArrayPool<KeyValuePair<string, object?>>.Shared.Rent(32);
+            int length = array.Length;
+
+            try
+            {
+                foreach (KeyValuePair<string, object?> item in tags)
+                {
+                    if (count == length)
+                        Grow(ref array, ref length);
+
+                    array[count++] = item;
+                }
+
+                if (count == 0)
+                    return Instrument.EmptyTags;
+
+                KeyValuePair<string, object?>[] result = new KeyValuePair<string, object?>[count];
+                array.AsSpan().Slice(0, count).CopyTo(result.AsSpan());
+                return result;
+            }
+            finally
```

Comment:
`try-finally` isn't needed for the ArrayPool.
In case of failures the array will just be dropped (i.e. collected by GC), no harm to the pool.

Reply:
It will, but I've always guarded against that. Here, it's a small array, so it's not a huge issue, but I believe it's a good practice to always ensure it's returned properly. Is there any performance-related reason to avoid the `try/finally` here?

Reply:
I was curious and found [this comment](https://github.com/dotnet/runtime/issues/71249#issuecomment-1165597563) from @stephentoub. The complexity of adding it here is not a big issue, but if there could be a small perf hit on the success path, that might be a reason to avoid it. We don't expect exceptions on this code path, so I can probably remove the `try/finally` here.

Reply:
@stevejgordon let's remove the `try-catch`.

Reply:
@tarekgh Removed!